contracts/BaseWrapper.sol

```diff
@@ -0,0 +1,162 @@
+// SPDX-License-Identifier: GPL-3.0
+pragma solidity ^0.6.12;
+pragma experimental ABIEncoderV2;
+
+import {ERC20} from "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+import {SafeMath} from "@openzeppelin/contracts/math/SafeMath.sol";
+
+import {RegistryAPI, VaultAPI} from "./BaseStrategy.sol";
+
+abstract contract BaseWrapper {
+    using SafeMath for uint256;
+
+    ERC20 public token;
+
+    // Reduce number of external calls (SLOADs stay the same)
+    VaultAPI[] private _cachedVaults;
+
+    RegistryAPI public registry;
+
+    constructor(address _token) public {
+        token = ERC20(_token);
+        // v2.registry.ychad.eth
+        registry = RegistryAPI(0xE15461B18EE31b7379019Dc523231C57d1Cbc18c);
+    }
+
+    function _setRegistry(address _registry) internal {
+        // In case you want to override the registry instead of re-deploying
+        registry = RegistryAPI(_registry);
+    }
+
+    function bestVault() public virtual view returns (VaultAPI) {
+        return VaultAPI(registry.latestVault(address(token)));
+    }
+
+    function allVaults() public virtual view returns (VaultAPI[] memory) {
+        uint256 cache_length = _cachedVaults.length;
+        uint256 num_deployments = registry.nextDeployment(address(token));
+
+        // Use cached
+        if (cache_length == num_deployments) {
+            return _cachedVaults;
+        }
+
+        VaultAPI[] memory vaults = new VaultAPI[](num_deployments);
+
+        for (uint256 deployment_id = 0; deployment_id < cache_length; deployment_id++) {
+            vaults[deployment_id] = _cachedVaults[deployment_id];
+        }
+
+        for (uint256 deployment_id = cache_length; deployment_id < num_deployments; deployment_id++) {
+            vaults[deployment_id] = VaultAPI(registry.vaults(address(token), deployment_id));
+        }
+
+        return vaults;
+    }
+
+    function _updateVaultCache(VaultAPI[] memory vaults) internal {
```

Comment:
thinking if there is a way to warm up the cache from a keeper from time to time and check if deployments length are different from cached we can update cache before withdraws and avoid users paying for the reorg in withdraw

Reply:
Could explore later. Worst case is one release needs to be cached by the user.